precompiles/prototype/prototype.go

```diff
@@ -0,0 +1,260 @@
+}
+
+// Abi() is required to implement the PrecompiledContract interface.
+func (c *Contract) Abi() abi.ABI {
+	return ABI
+}
+
+// RequiredGas is required to implement the PrecompiledContract interface.
+// The gas has to be calculated deterministically based on the input.
+func (c *Contract) RequiredGas(input []byte) uint64 {
+	// base cost to prevent large input size
+	baseCost := uint64(len(input)) * c.kvGasConfig.WriteCostPerByte
+
+	// get methodID (first 4 bytes)
+	var methodID [4]byte
+	copy(methodID[:], input[:4])
+
+	if requiredGas, ok := GasRequiredByMethod[methodID]; ok {
+		return requiredGas + baseCost
+	}
+
+	return baseCost
+}
+
+func (c *Contract) Bech32ToHexAddr(method *abi.Method, args []interface{}) ([]byte, error) {
+	if len(args) != 1 {
+		return nil, &ptypes.ErrInvalidNumberOfArgs{
+			Got:    len(args),
+			Expect: 1,
+		}
+	}
+
+	address, ok := args[0].(string)
+	if !ok || address == "" {
+		return nil, fmt.Errorf("invalid bech32 address: %v", args[0])
+	}
+
+	bech32Prefix := strings.SplitN(address, "1", 2)[0]
+	if bech32Prefix == address {
+		return nil, fmt.Errorf("invalid bech32 address: %s", address)
+	}
+
+	addressBz, err := sdk.GetFromBech32(address, bech32Prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := sdk.VerifyAddressFormat(addressBz); err != nil {
+		return nil, err
+	}
+
+	return method.Outputs.Pack(common.BytesToAddress(addressBz))
+}
+
+func (c *Contract) Bech32ify(method *abi.Method, args []interface{}) ([]byte, error) {
+	if len(args) != 2 {
+		return nil, &ptypes.ErrInvalidNumberOfArgs{
+			Got:    len(args),
+			Expect: 2,
+		}
+	}
+
+	cfg := sdk.GetConfig()
+	prefix, _ := args[0].(string)
+	if strings.TrimSpace(prefix) == "" {
+		return nil, fmt.Errorf(
+			"invalid bech32 human readable prefix (HRP). Please provide a either an account, validator or consensus address prefix (eg: %s, %s, %s)",
+			cfg.GetBech32AccountAddrPrefix(),
+			cfg.GetBech32ValidatorAddrPrefix(),
+			cfg.GetBech32ConsensusAddrPrefix(),
+		)
+	}
+
+	address, ok := args[1].(common.Address)
+	if !ok {
+		return nil, fmt.Errorf("invalid hex address")
+	}
+
+	// NOTE: safety check, should not happen given that the address is is 20 bytes.
+	if err := sdk.VerifyAddressFormat(address.Bytes()); err != nil {
+		return nil, err
+	}
+
+	bech32Str, err := sdk.Bech32ifyAddressBytes(prefix, address.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	addressBz, err := sdk.GetFromBech32(bech32Str, "zeta")
+	if err != nil {
+		return nil, err
+	}
+
+	if err := sdk.VerifyAddressFormat(addressBz); err != nil {
+		return nil, err
+	}
+
+	return method.Outputs.Pack(bech32Str)
+}
+
+func (c *Contract) GetGasStabilityPoolBalance(
+	ctx sdk.Context,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	if len(args) != 1 {
+		return nil, &(ptypes.ErrInvalidNumberOfArgs{
+			Got:    len(args),
+			Expect: 1,
+		})
+	}
+
+	// Unwrap arguments. The chainID is the first and unique argument.
+	chainID, ok := args[0].(int64)
```

Comment:
Initially I thought so, but changed it because of GetGasStabilityPoolBalance, which signature is:
`GetGasStabilityPoolBalance(ctx sdk.Context, chainID int64)`

This way seems to have less friction as we expect an int64 and it's passed directly to the function, without conversion. or possibility of overflow.